From 52b9939d07b0ca03f8a37ec5cb2f1bd45c7d799a Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Thu, 25 Jan 2024 11:13:23 +0200
Subject: [PATCH] [SW-172027] gaudi2: security: modify svn version check

Due to bug in arcboot rom, firmware svn version one less than
the efuse svn version was allowed to boot, ex: if efuse svn = 7,
firmware svn 6 was allowed to boot. To fix this it is decided
to program an additional efuse bit in SVN. i,e. if SVN 7 then efuse
bits 0 to 7 are programmed (0xFF). This patch adds support to decrement
svn version read from efuse by 1.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I3256192629294bdbb416d124b1877d2d072874e9
---

diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index a2695ef..bb27b00 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -587,6 +587,14 @@
 	/* Add bits counted by iterator in case that all the bits are set */
 	svn += i * 32;
 
+	/*
+	 * SVN was computed by counting no of set bits, however due to bug in arcboot rom an
+	 * addtional bit is programmed in efuse hence need to subract the svn by 1 bit.
+	 * ex: efuse val = 0xFF when SVN = 7, efuse val = 0x7F when SVN = 6
+	 */
+	if (svn)
+		svn--;
+
 	preboot_puts_hex("SVN on eFuse: ", svn, 0);
 	return svn;
 }
